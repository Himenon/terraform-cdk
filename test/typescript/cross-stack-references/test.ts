import * as fs from "fs-extra";
import * as path from "path";
import { TestDriver } from "../../test-helper";

describe("cross stack references", () => {
  let driver: TestDriver;

  function readLocalFile(fileName: string): string {
    return fs.readFileSync(
      path.join(driver.workingDirectory, fileName),
      "utf8"
    );
  }

  beforeAll(async () => {
    driver = new TestDriver(__dirname);
    await driver.setupTypescriptProject();
    console.log(driver.workingDirectory);
    await driver.synth();
  });

  test("synth generates JSON", () => {
    expect(driver.synthesizedStack("source").toString()).toMatchSnapshot();
    expect(driver.synthesizedStack("passthrough").toString()).toMatchSnapshot();
    // expect(driver.synthesizedStack("sink").toString()).toMatchSnapshot();

    expect(driver.manifest()).toMatchSnapshot();
  });

  describe("deployed", () => {
    beforeAll(async () => {
      await driver.deploy("source");
      await driver.deploy("passthrough");
      // await driver.deploy("sink");
    });

    it("prereq - source stack writes files", () => {
      expect(readLocalFile("originNum")).toMatchInlineSnapshot(`"4"`);
      expect(readLocalFile("originStr")).toMatchInlineSnapshot(
        `"BN@LPd]x*VqopF[B5#2>?}_WbbJ*Vm{S"`
      );
    });

    it.skip("references primitive values", () => {
      expect(readLocalFile("originNum")).toBe(readLocalFile("passthroughNum"));
      expect(readLocalFile("originStr")).toBe(readLocalFile("passthroughStr"));
    });

    it.todo("references can be passed through stacks");
    it.todo("can use reference in terraform function");

    it.todo("references terraform function output");

    it.todo("references complex values");
    it.todo("references nested values");
  });
});
