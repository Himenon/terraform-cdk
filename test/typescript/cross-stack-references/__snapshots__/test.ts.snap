// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cross stack references synth generates JSON 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"source\\",
      \\"backend\\": \\"local\\"
    }
  },
  \\"terraform\\": {
    \\"required_providers\\": {
      \\"random\\": {
        \\"version\\": \\"= 3.1.0\\",
        \\"source\\": \\"hashicorp/random\\"
      },
      \\"local\\": {
        \\"version\\": \\"=2.1.0\\",
        \\"source\\": \\"hashicorp/local\\"
      }
    },
    \\"backend\\": {
      \\"local\\": {}
    }
  },
  \\"provider\\": {
    \\"random\\": [
      {}
    ],
    \\"local\\": [
      {}
    ]
  },
  \\"resource\\": {
    \\"random_integer\\": {
      \\"int\\": {
        \\"max\\": 6,
        \\"min\\": 0,
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"source/int\\",
            \\"uniqueId\\": \\"int\\"
          }
        }
      }
    },
    \\"random_password\\": {
      \\"str\\": {
        \\"length\\": 32,
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"source/str\\",
            \\"uniqueId\\": \\"str\\"
          }
        }
      }
    },
    \\"local_file\\": {
      \\"originNum\\": {
        \\"content\\": \\"\${random_integer.int.result}\\",
        \\"filename\\": \\"../../../originNum\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"source/originNum\\",
            \\"uniqueId\\": \\"originNum\\"
          }
        }
      },
      \\"originStr\\": {
        \\"content\\": \\"\${random_password.str.result}\\",
        \\"filename\\": \\"../../../originStr\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"source/originStr\\",
            \\"uniqueId\\": \\"originStr\\"
          }
        }
      }
    }
  },
  \\"output\\": {
    \\"crossstackoutputrandomintegerintresult\\": {
      \\"value\\": \\"\${random_integer.int.result}\\",
      \\"sensitive\\": true
    },
    \\"crossstackoutputrandompasswordstrresult\\": {
      \\"value\\": \\"\${random_password.str.result}\\",
      \\"sensitive\\": true
    }
  }
}"
`;

exports[`cross stack references synth generates JSON 2`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"passthrough\\",
      \\"backend\\": \\"local\\"
    }
  },
  \\"terraform\\": {
    \\"required_providers\\": {
      \\"random\\": {
        \\"version\\": \\"= 3.1.0\\",
        \\"source\\": \\"hashicorp/random\\"
      },
      \\"local\\": {
        \\"version\\": \\"=2.1.0\\",
        \\"source\\": \\"hashicorp/local\\"
      }
    }
  },
  \\"provider\\": {
    \\"random\\": [
      {}
    ],
    \\"local\\": [
      {}
    ]
  },
  \\"resource\\": {
    \\"local_file\\": {
      \\"passthroughNum\\": {
        \\"content\\": \\"\${\${data.terraform_remote_state.crossstackreferenceinputsource.outputs.crossstackoutputrandomintegerintresult}}\\",
        \\"filename\\": \\"../../../passthroughNum\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"passthrough/passthroughNum\\",
            \\"uniqueId\\": \\"passthroughNum\\"
          }
        }
      },
      \\"passthroughStr\\": {
        \\"content\\": \\"\${\${data.terraform_remote_state.crossstackreferenceinputsource.outputs.crossstackoutputrandompasswordstrresult}}\\",
        \\"filename\\": \\"../../../passthroughStr\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"passthrough/passthroughStr\\",
            \\"uniqueId\\": \\"passthroughStr\\"
          }
        }
      }
    }
  },
  \\"data\\": {
    \\"terraform_remote_state\\": {
      \\"crossstackreferenceinputsource\\": {
        \\"backend\\": \\"local\\",
        \\"config\\": {
          \\"path\\": \\"/tmp/testE4CWte/terraform.source.tfstate\\"
        }
      }
    }
  }
}"
`;

exports[`cross stack references synth generates JSON 3`] = `
"{
  \\"version\\": \\"stubbed\\",
  \\"stacks\\": {
    \\"source\\": {
      \\"name\\": \\"source\\",
      \\"constructPath\\": \\"source\\",
      \\"workingDirectory\\": \\"stacks/source\\",
      \\"synthesizedStackPath\\": \\"stacks/source/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": []
    },
    \\"passthrough\\": {
      \\"name\\": \\"passthrough\\",
      \\"constructPath\\": \\"passthrough\\",
      \\"workingDirectory\\": \\"stacks/passthrough\\",
      \\"synthesizedStackPath\\": \\"stacks/passthrough/cdk.tf.json\\",
      \\"annotations\\": [],
      \\"dependencies\\": [
        \\"source\\"
      ]
    }
  }
}"
`;
